### ISSUE TITLE
Input method support on Wayland

### ISSUE THREAD

**Issue by mykhail0** [qtile/qtile#4623]
### Issue description

Input method functionality (fcitx5, ibus) doesn't work fully on Wayland backend. I was not able to make ibus work at all, and fcitx5 works only in GTK/Qt based applications.

fcitx5 is not usable in terminal apps such as alacritty or kitty unless they are run under XWayland, e.g. ```$ env WAYLAND_DISPLAY= alacritty```.

Steps to reproduce:

1. Install fcitx5
2. configure it with fcitx5-configtool and add some input method (e.g. Keyboard - Ukrainian, to see immediately if Cyrillic alphabet is being typed)
3. In Global Options under Enumerate Input Method Forward there is a shortcut for changing input method
4. ```$ fcitx5 & ; env WAYLAND_DISPLAY= alacritty```, change input method and confirm fcitx5 works by typing Cyrillic
5. Now witness no effects when alacritty is run normally

To see absence of popup box for Chinese method you need to additionally install appropriate package with fcitx5 pinyin input method for your distribution (fcitx5-chinese-addons for Arch). Then you will be able to add pinyin input method in fcitx5-configtool.

The issue is Qtile specific, since it is not present on hyprland (from what I gathered on Github).
Swaywm had a similar issue, which is somewhat resolved now, since one can use fcitx5 but not with input methods requiring popup selection screen (e.g. Mandarin Chinese) (see https://github.com/alacritty/alacritty/issues/5223), because there is no popup screen. This is being further resolved in https://github.com/swaywm/sway/pull/7226.

There are 3 protocols with Wayland for input methods:
https://wayland.app/protocols/input-method-unstable-v1
https://github.com/swaywm/wlroots/blob/master/protocol/input-method-unstable-v2.xml
https://wayland.app/protocols/text-input-unstable-v3

fcitx5 and swaywm both implement v2 and v3. ibus seems to implement v1. Kitty implements v3, not sure about alacritty.

Linux distribution I'm using is Arch.
My ```/etc/profile``` settings for getting fcitx5 to work with GTK/Qt based apps (as far as I know, no variables are necessary for alacritty):
```shell
export QT_QPA_PLATFORM="wayland;xcb"
export QT_IM_MODULE=fcitx
export QT_SELECT=6
export GTK_IM_MODULE=fcitx
export XMODIFIERS="@im=fcitx"
export SDL_IM_MODULE=fcitx
```

### Version

0.23.0

### Backend

Wayland (experimental)

### Required

- [X] I have searched past issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by moralrebuild** [comment]
any update?

**Comment by iacore** [comment]
ibus definitely doesn't work under Qtile+wayland.

**Comment by eternal-sorrow** [comment]
> ibus definitely doesn't work under Qtile+wayland.

To be fair, ibus doesn't work under sway either. But fcitx does.

**Comment by iacore** [comment]
So what does QTile implement?

**Comment by okphilo** [comment]
why no one care this issue? CJK input method can't use in qtile under wayland, in qt and gtk app, im-module works , but for other apps (like terminal) , can't use input method.
qtile does not support wayland input method protocol, and there is not plan to support it, right?

**Comment by iacore** [comment]
wayland's support for IME is :)

**Comment by okphilo** [comment]
> wayland's support for IME is :)

no, it's not wayland issue, gnome/kde/sway/lxqt....... all of this is good, it's qtile issue.

**Comment by TheHighGroundWins** [comment]
Just switched to Wayland and like the improved display experience, but it sucks to not be able to type in multilingual input on terminal.

**Comment by TheHighGroundWins** [comment]
As a temporary solution, for none IME based languages changing the keyboard layout and variant works. Although it is recommended to have some kind of shortcut or have typed the same thing for english, so that you don't get stuck in a non latin based language.

You can do so by typing: qtile cmd-obj -o core -f set_keymap -a [layout] [options] [variant]
For example: qtile cmd-obj -o core -f set_keymap -a "mn" for just layout.

Currently I am trying to find the matching pywayland protocol for the multilingual protocols. As they have slightly different names. Then later figure out how to implement it via qtile built in functions.

**Comment by TheHighGroundWins** [comment]
Input method v3 is currently unsupported by pywlroots and cannot be implemented without adding to pywlroots.

However, with the codebase being rewritten in C it would be better to wait until that is done so that input method v3 can be implemented without having to add to pywlroots.

**Comment by eternal-sorrow** [comment]
Ibus only recently implemented v2 https://github.com/ibus/ibus/pull/2256

**Comment by eternal-sorrow** [comment]
This is the only issue that prevents me from migration from sway to qtile.

**Comment by eternal-sorrow** [comment]
Is this "rewrite in C" only hypothetical at this point or there are steps being taken in that direction?

**Comment by TheHighGroundWins** [comment]
> Is this "rewrite in C" only hypothetical at this point or there are steps being taken in that direction?

It's a real thing that is confirmed to be happening and most likely will be done around April or May.

@jwijenbergh is doing this rewrite, based on other Wayland compositors.

**Comment by aaannestad** [comment]
Is there any movement on this since that last update? I'm also in the camp of 'this is the only thing preventing me from using qtile'.

**Comment by TheHighGroundWins** [comment]
> Is there any movement on this since that last update? I'm also in the camp of 'this is the only thing preventing me from using qtile'.

Unfortunately no, as the C rewrite of qtile that is supposed allow the implementation of the protocol is quite and undertaking for the one person working on it and doesn't have a defined ETA. Even if the rewrite is done implementing the protocol won't be easy.

I have personally switched to hyprland, and customized it to be as close to qtile, as IME as even essential for me that I can't compromise on.

Wayland in general is quite rough on qtile which is why is labeled experimental due to how its added.

**Comment by eternal-sorrow** [comment]
Not stale. Automated issue closing shouldn't be a thing.

**Comment by eternal-sorrow** [comment]
Now that C backend is released, I hope there will be movement in this direction.

**Comment by jwijenbergh** [comment]
There definitely will be!

**Comment by matlinuxer2** [comment]
According to the post's survey: https://zhuanlan.zhihu.com/p/22611314767 ,
Adding Wayland's text-input-v3[1] could be the best route to support CJK input methods for most cases.
Chromium also supported text-input-v3, which is followed by Chrome, Electron, and all Electron-based apps. [2]

[1]. https://wayland.app/protocols/text-input-unstable-v3
[2]. https://issues.chromium.org/issues/40113488
